enhancements/catalog-source-config.md

```diff
@@ -0,0 +1,340 @@
+
+```yaml
+spec:
+  type: object
+  required:
+    - sourceConfig
+  properties:
+    # Deprecate 
+    # image: ... 
+    # address: ... 
+    # configMap: ...
+    # sourceType: grpc | configMap | internal
+    
+    # Keep as is:
+    description: ...
+    displayName: ...
+    icon: ...
+    priority: ...
+    publisher: ...
+    secrets: ...
+
+    # Add:
+    # Note: 'internal' type is being deprecated so it gets no config structure
+    grpcPodSource:
+      podSpec:  # a subset (or the whole of) corev1.PodSpec
+        image: ... # required
+        # all other attributes are optional
+        # and override the default catalog source pod spec
+        ...
+      ...
+    grpcServiceSource:
+      address: ...
+      ...
+    configMapSource:
+      configMap: ...
+      ... 
+  oneOf:
+    - required: ["grpcPodSource"]
+    - required: ["grpcServiceSource"]
+    - required: ["configMapSource"]
+                  
+``` 
+
+This option introduces breaking changes. This means that while we are on the depracation path, the users will be able to configure their `CatalogSource`s in the current way, _OR_, use one of `grpcPodSource`, `grpcServiceSource`, or `configMapSource` (with precendence given to the new configuration). 
+
+**Pros**
+  1. Each CatalogSource type has its own configuration structure
+  2. We don't polute the top level with endless optional attributes that behave differently depending on the values of other attributes
+  3. Better UX
+  4. New catalog source types can be added easily
+
+
+**Cons**
+  1. Introduces breaking changes
+  2. There is not first-class support for union type validation in kubebuilder/operator-sdk. We'll need another mechanism to add and test these validations (e.g. patch that gets executed after `make controller-gen` in the `api` repo)
+
+#### CatalogSourceClass
+
+In this option, we transion the `CatalogSource` to a more plugable architecture by introducing a `catalogSourceClass` and a schemaless `parameters` block where arbitrary parameters can be passed in. The controller would then be in charge of reconciling catalog sources of classes it recognizes.
+
+In the future, we'd have something like this:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha2
+kind: CatalogSource
+metadata:
+  name: operatorhubio-catalog
+  namespace: olm
+spec:
+  catalogSourceClass: grpc-pod
+  displayName: Community Operators
+  publisher: OperatorHub.io
+  parameters:
+    image: quay.io/operatorhubio/catalog:latest
+    # additional parameters are options and override the underlying podspec
+```
+
+**Pros**
+  1. Extensible architecture: users can provide their own `CatalogSourceClass` types and controllers
+  2. Better UX than the status quo: top level attributes are shared and `parameters` should only contain information relevant for that class of catalog source
+
+**Cons**
+  1. `parameters` is schemaless. So, we'd also need an adminission webhook
+  2. Requires deprecation
+
```

Comment:
I would have expected the CatalogSourceClass to be a separate resource (containing the parameters), similar to storage classes. https://kubernetes.io/docs/concepts/storage/storage-classes/
An advantage of the approach is that you can have the field catalogSourceClass as optional defaulting to the current behaviour.  This is backward compatible and there is no further change to the CatalogSource.

Reply:
Yeah, that's a great call out. I think I got this all wrong. This is probably not a good approach then. I don't think having a CatalogSourceClass (the right way) would obviate the underlying problem here. We'd want to attach these to overrides to specifc CatalogSources.

Reply:
I've completely removed this option. Just didn't make sense anymore.